File: algorithms/1178. Number of Valid Words for Each Puzzle/solution.py
class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        counter = collections.Counter(''.join(sorted(set(w))) for w in words)
        answer = []
        for p in puzzles:
            cnt = 0
            for i in range(7):
                for c in itertools.combinations(p[1:], i):
                    cnt += counter[''.join(sorted((p[0],) + c))]
            answer.append(cnt)
        return answer

File: algorithms/1178. Number of Valid Words for Each Puzzle/solution.cc
class Solution {
public:
    vector<int> findNumOfValidWords(vector<string>& words, vector<string>& puzzles) {
        // learned from jiah: use a trick to generate combination
        unordered_map<int, int> maskCnt;
        for (auto & w : words) {
            ++maskCnt[getMask(w)];
        }
        vector<int> answer;
        for (auto & p : puzzles) {
            int mask = getMask(p);
            int initMask = 1 << (p[0] - 'a');
            int cnt = 0;
            for (int i = mask; i > 0; i = (i - 1) & mask) {
                if ((initMask & i) && maskCnt.count(i)) {
                    cnt += maskCnt[i];
                }
            }
            answer.push_back(cnt);
        }
        return answer;
    }

    int getMask(string & s) {
        int mask = 0;
        for (char c : s) {
            mask |= 1 << (c - 'a');
        }
        return mask;
    }
};
